sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala

```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
     DataFrameHolder(
       _sqlContext.internalCreateDataFrame(rows, StructType(StructField("_1", dataType) :: Nil)))
   }
+
+  /**
+   * ::Experimental::
+   *
+   * Pimp my library decorator for tungsten caching of DataFrames.
+   * @since 1.5.1
+   */
+  @Experimental
+  implicit class TungstenCache(df: DataFrame) {
+    /**
+     * Packs the rows of [[df]] into contiguous blocks of memory.
+     * @param compressionType "" (default), "lz4", "lzf", or "snappy", see
+     *                        [[CompressionCodec.ALL_COMPRESSION_CODECS]]
+     * @param blockSize size of each MemoryBlock (default = 4 MB)
+     */
+    def tungstenCache(
+        compressionType: String = "", blockSize: Int = 4000000): (RDD[_], DataFrame) = {
+      val schema = df.schema
+
+      val convert = CatalystTypeConverters.createToCatalystConverter(schema)
+      val internalRows = df.rdd.map(convert(_).asInstanceOf[InternalRow])
+      val cachedRDD = internalRows.mapPartitions { rowIterator =>
+        val bufferedRowIterator = rowIterator.buffered
+        val convertToUnsafe = UnsafeProjection.create(schema)
+        val taskMemoryManager = new TaskMemoryManager(SparkEnv.get.executorMemoryManager)
+        val compressionCodec: Option[CompressionCodec] = if (compressionType.isEmpty) {
+          None
+        } else {
+          Some(CompressionCodec.createCodec(SparkEnv.get.conf, compressionType))
+        }
+
+        new Iterator[MemoryBlock] {
+          // NOTE: This assumes that size of every row < blockSize
+          def next(): MemoryBlock = {
+            // Packs rows into a `blockSize` bytes contiguous block of memory, starting a new block
+            // whenever the current fills up
+            // Each row is laid out in memory as [rowSize (4)|row (rowSize)]
+            val block = taskMemoryManager.allocateUnchecked(blockSize)
+
+            var currOffset = 0
+            val seenRows = ArrayBuffer[InternalRow]()
+            while (bufferedRowIterator.hasNext && currOffset < blockSize) {
+              val currRow = convertToUnsafe.apply(bufferedRowIterator.head)
+              val recordSize = 4 + currRow.getSizeInBytes
+              if (currOffset + recordSize < blockSize) {
+                Platform.putInt(
+                  block.getBaseObject, block.getBaseOffset + currOffset, currRow.getSizeInBytes)
+                currRow.writeToMemory(block.getBaseObject, block.getBaseOffset + currOffset + 4)
+                seenRows += currRow
```

Comment:
Don't we need to perform a defensive copy here if we're going to buffer these unsafe rows, since the row returned from the projection will be backed by a re-used buffer inside of that projection?